File: ScriptsToProcess/INITKEYS.py
"""
BIM KEY INITIALIZER (RESTORED & CORE-FREE)
================================================================================
Injects the required Attribute User Text keys into referenced Rhino objects.
* FIXED: Restored Rhino API injection logic and GuidsOut pass-through.
* UPDATED: Core mechanics (Type, CoreHeight) completely removed.
* UPDATED: Added <null> data stream interceptors.

INPUTS:
    Guids (Guid) [List Access] : Referenced Rhino geometry IDs.
    Run   (bool) [Item Access] : Wire a Button here to execute the injection.

OUTPUTS:
    GuidsOut (Guid) : Pass-through for the Serializer.
================================================================================
"""

import Rhino
import System
import time

# --- UI WRAPPER ---
def update_ui(msg):
    _ghenv = globals().get('ghenv')
    _comp = globals().get('component')
    if _ghenv: _ghenv.Component.Message = msg
    elif _comp: _comp.Message = msg
    else: print(msg)

_ghenv = globals().get('ghenv')
if _ghenv:
    _ghenv.Component.Name = "BIM Key Initializer"
    _ghenv.Component.NickName = "INITKEYS"
    _ghenv.Component.Description = "Safely injects default BIM attributes into referenced Rhino curves."

# ==============================================================================
# MAIN EXECUTION
# ==============================================================================
def initialize_bim_keys():
    exec_start = time.perf_counter()
    
    guids_in = globals().get('Guids', [])
    run_btn = globals().get('Run', False)
    
    status_msg = []
    
    # Intercept nulls from empty branches or dead wires
    clean_guids = [g for g in guids_in if g is not None]
    
    if not clean_guids:
        status_msg = ["Error: No Curves Connected"]
        clean_guids = []
    elif not run_btn:
        status_msg = ["Ready.", "Press Button."]
    else:
        doc = Rhino.RhinoDoc.ActiveDoc
        objects_modified = 0
        invalid_objects = 0
        
        # The new stripped-down schema (No Core logic)
        default_schema = {
            "BuildingID": "Building_01",
            "TowerID": "Main_Tower", 
            "Program": "Residential",
            "Phase": "0",
            "Floors": "1",
            "FloorHeight": "4.0"
        }
        
        for item in clean_guids:
            if not isinstance(item, System.Guid):
                invalid_objects += 1
                continue
                
            try:
                obj = doc.Objects.FindId(item)
                if not obj:
                    invalid_objects += 1
                    continue
            except TypeError:
                invalid_objects += 1
                continue
                
            modified = False
            
            for key, default_val in default_schema.items():
                existing_val = obj.Attributes.GetUserString(key)
                if existing_val is None:
                    obj.Attributes.SetUserString(key, default_val)
                    modified = True
                    
            if modified:
                obj.CommitChanges()
                objects_modified += 1
                
        if invalid_objects > 0 and objects_modified == 0:
            status_msg = [f"Failed: {invalid_objects} unreferenced", "(Check Type Hint = Guid)"]
        elif invalid_objects > 0:
            status_msg = [f"Injected: {objects_modified}", f"Skipped: {invalid_objects} (Not ref'd)"]
        else:
            status_msg = [f"Success: {objects_modified} objects"]
            
    exec_time = (time.perf_counter() - exec_start) * 1000
    
    # Standardized UI layout
    msg_lines = [
        "INITKEYS",
        f"Time: {exec_time:.0f} ms",
        "---"
    ] + status_msg
    
    update_ui("\n".join(msg_lines))
    return clean_guids

# Outputs
GuidsOut = initialize_bim_keys()

File: Components/InitKeys.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using Grasshopper.Kernel;
using Rhino;
using Rhino.DocObjects;

namespace Enzyme.Components
{
    public class InitKeysComponent : GH_Component
    {
        public InitKeysComponent()
          : base("BIM Key Initializer", "INITKEYS",
              "Safely injects default BIM attributes into referenced Rhino curves.",
              Enzyme.Utils.TabInfo.TabName, "Masterplan (Beta)")
        {
        }

                public override void AddedToDocument(GH_Document document)
        {
            base.AddedToDocument(document);
            if (this.Attributes == null) this.CreateAttributes();

            bool hasSources = false;
            foreach (var param in this.Params.Input)
                if (param.SourceCount > 0) { hasSources = true; break; }

            if (!hasSources)
            {
                Enzyme.Utils.AutoWireHelper.WireButton(this, document, 1, 210, 0);
            }
        }

        protected override void RegisterInputParams(GH_Component.GH_InputParamManager pManager)
        {
            pManager.AddGenericParameter("Guids", "G", "Referenced Rhino geometry IDs.", GH_ParamAccess.list);
            pManager.AddBooleanParameter("Run", "R", "Wire a Button here to execute the injection.", GH_ParamAccess.item);
            pManager[0].Optional = true;
            pManager[1].Optional = true;
        }

        protected override void RegisterOutputParams(GH_Component.GH_OutputParamManager pManager)
        {
            pManager.AddGenericParameter("Guids", "G", "Pass-through for the Serializer.", GH_ParamAccess.list);
        }

        public override GH_Exposure Exposure => GH_Exposure.primary;

        protected override void SolveInstance(IGH_DataAccess DA)
        {
            Stopwatch sw = Stopwatch.StartNew();

            List<Grasshopper.Kernel.Types.IGH_Goo> guidsInGoo = new List<Grasshopper.Kernel.Types.IGH_Goo>();
            DA.GetDataList(0, guidsInGoo);

            bool runBtn = false;
            DA.GetData(1, ref runBtn);

            List<string> statusMsg = new List<string>();
            List<Guid> cleanGuids = new List<Guid>();

            foreach (var goo in guidsInGoo)
            {
                if (goo == null) continue;
                
                Guid g = Guid.Empty;
                if (goo.CastTo<Guid>(out g))
                {
                    if (g != Guid.Empty)
                        cleanGuids.Add(g);
                }
                else if (goo is Grasshopper.Kernel.Types.GH_String ghStr)
                {
                    if (Guid.TryParse(ghStr.Value, out Guid parsedG) && parsedG != Guid.Empty)
                        cleanGuids.Add(parsedG);
                }
                else if (goo is Grasshopper.Kernel.Types.GH_Guid ghGuid)
                {
                    if (ghGuid.Value != Guid.Empty)
                        cleanGuids.Add(ghGuid.Value);
                }
            }

            if (cleanGuids.Count == 0)
            {
                statusMsg.Add("Error: No Curves Connected");
            }
            else if (!runBtn)
            {
                statusMsg.Add("Ready.");
                statusMsg.Add("Press Button.");
            }
            else
            {
                RhinoDoc doc = RhinoDoc.ActiveDoc;
                int objectsModified = 0;
                int invalidObjects = 0;

                var defaultSchema = new Dictionary<string, string>
                {
                    { "BuildingID", "Building_01" },
                    { "TowerID", "Main_Tower" },
                    { "Program", "Residential" },
                    { "Phase", "0" },
                    { "Floors", "1" },
                    { "FloorHeight", "4.0" }
                };

                foreach (Guid item in cleanGuids)
                {
                    RhinoObject obj = doc?.Objects.FindId(item);
                    if (obj == null)
                    {
                        invalidObjects++;
                        continue;
                    }

                    bool modified = false;

                    foreach (var kvp in defaultSchema)
                    {
                        string existingVal = obj.Attributes.GetUserString(kvp.Key);
                        if (existingVal == null)
                        {
                            obj.Attributes.SetUserString(kvp.Key, kvp.Value);
                            modified = true;
                        }
                    }

                    if (modified)
                    {
                        obj.CommitChanges();
                        objectsModified++;
                    }
                }

                if (invalidObjects > 0 && objectsModified == 0)
                {
                    statusMsg.Add($"Failed: {invalidObjects} unreferenced");
                    statusMsg.Add("(Check Type Hint = Guid)");
                }
                else if (invalidObjects > 0)
                {
                    statusMsg.Add($"Injected: {objectsModified}");
                    statusMsg.Add($"Skipped: {invalidObjects} (Not ref'd)");
                }
                else
                {
                    statusMsg.Add($"Success: {objectsModified} objects");
                }
            }

            sw.Stop();
            long execTime = sw.ElapsedMilliseconds;

            string finalMsg = $"{this.NickName}\nTime: {execTime} ms\n---\n" + string.Join("\n", statusMsg);
            this.Message = finalMsg;

            DA.SetDataList(0, cleanGuids);
        }

        protected override System.Drawing.Bitmap Icon
        {
            get
            {
                return IconLoader.Load("INITKEYS.png");
            }
        }

        public override Guid ComponentGuid
        {
            get { return new Guid("14d6ea4e-1234-4bc9-93af-e0dc2857053e"); }
        }
    }
}
